weasyl/controllers/moderation.py

```diff
@@ -219,3 +219,201 @@ def modcontrol_copynotetostaffnotes_post_(request):
+            comment_content=content,
+        )
+        welcome.journal_insert(userid=userid, journalid=journalid, rating=rating, settings=settings)
+    elif action == "delete":
+        # Delete and purge the journal from the welcome table
+        define.engine.execute("""
+            DELETE FROM journal
+            WHERE journalid = %(id)s AND
+                  is_spam = TRUE
+        """, id=journalid)
+        welcome.journal_remove(journalid=journalid)
+    else:
+        raise HTTPSeeOther("/modcontrol/spamqueue/journal")
+
+    raise HTTPSeeOther("/modcontrol/spamqueue/journal")
+
+
+@moderator_only
+def modcontrol_spamqueue_submission_get_(request):
+    """
+    A Pyramid route controller for displaying any journal items which the spam filter has tagged as being spam.
+    :param request: The Pyramid request.
+    :return: A rendered page for the journal spam queue.
+    """
+    query = define.engine.execute("""
+        SELECT lo.login_name, su.submitid, su.title, su.content
+        FROM login lo
+        INNER JOIN submission su USING (userid)
+        WHERE su.is_spam IS TRUE
+    """).fetchall()
+    return Response(define.webpage(
+        request.userid,
+        "modcontrol/spamqueue/submission.html",
+        [query],
+        title="Submission Spam Queue"
+    ))
+
+
+@moderator_only
+@token_checked
+def modcontrol_spamqueue_submission_post_(request):
+    """
+    A Pyramid route controller for approving for display or items which the spam filter has tagged as being spam.
+    :param request: The Pyramid request.
+    :return/raises: HTTPSeeOther for the journal spam queue.
+    """
+    directive = request.params.get("directive")
+    if directive:
+        submitid, action = directive.split(",")
+    else:
+        raise HTTPSeeOther("/modcontrol/spamqueue/journal")
+
+    if action == "approve":
+        # Approve and insert the journal into the notifications table.
+        submitid, userid, rating, settings, content, ua_id, ip_addr = define.engine.execute("""
+            UPDATE submission
+            SET is_spam = FALSE
+            WHERE submitid = %(id)s
+            RETURNING submitid, userid, rating, settings, content, submitter_user_agent_id, submitter_ip_address
+        """, id=submitid).first()
+        # Update the spam filtering backend to indicate that this was not spam
+        spam_filtering.submit(
+            is_ham=True,
+            user_ip=ip_addr,
+            user_agent_id=ua_id,
+            user_id=userid,
+            comment_type="submission",
+            comment_content=content,
+        )
+        welcome.submission_insert(userid=userid, submitid=submitid, rating=rating, settings=settings)
+    elif action == "delete":
+        # Delete and purge the submission from the welcome table
+        define.engine.execute("""
+            DELETE FROM submission
+            WHERE submitid = %(id)s AND
+                  is_spam = TRUE
+        """, id=submitid)
```

Comment:
The more I think on this, the more I dislike directly deleting ``submission`` records.